### ISSUE TITLE
`:discouraged-var`: deem clojure.core as equivalent to cljs.core?

### ISSUE THREAD

**Issue by vemv** [clj-kondo/clj-kondo#2169]
I tried the following `:discouraged-var` rule, in a cljs-oriented project:

```clj
clojure.core/for {:message "Laziness can cause rendering issues"}
```

It didn't work for `clojure.core/for`, while specifying it for `cljs.core/for` did.

I would say, for cljs users, clojure.core and cljs.core are largely equivalent.

In specific terms, all of these work in a .cljs file:

```clj
(for ,,,)
(clojure.core/for ,,,)
(cljs.core/for ,,,)
```

So, ideally I could specify `clojure.core/for` and get what I meant. Otherwise there's a risk of a false negative.

WDYT?

Thanks - V

**Comment by borkdude** [comment]
I admit that this is a potential source of confusion, but note that this works the same for all linters, not only for :discouraged-var.

**Comment by borkdude** [comment]
Yeah, maybe we could just allow non-qualified symbols, like `for` to no break people who rely on the already-in-place behavior.

**Comment by vemv** [comment]
> Yeah, maybe we could just allow non-qualified symbols, like for to not break people who rely on the already-in-place behavior.

I thought about that, however, from the user perspective, `for {:message "Laziness can cause rendering issues"}` might look ambiguous - does it refer only to clj/s core `for`, or to any `for` that happened to be `:refer`ed into the given namespace?

---

I have a consideration to add: cljs.core and clojure.core can only considered equivalent _on a cljs context_. e.g., if kondo is linting a .clj file, a rule defined for `cljs.core/for` should be irrelevant.

So things would asymmetric (just like there's an asymmetry when comparing the clj and cljs compilers).

In face of that consideration, does the following concern still apply...? 

> Yeah, maybe we could just allow non-qualified symbols, like for to not break people who rely on the already-in-place behavior.

As I imagine it, a useful existing pattern (worth not 'breaking') would not be possible. I'd be open to change my mind for a specific example.

**Comment by borkdude** [comment]
Another option could be to have one symbol to indicate all clojure dialects: `core/+` which would mean `clojure.core/+` and `cljs.core/+`. You can accomplish this in user space by introducing a namespace group:

https://github.com/clj-kondo/clj-kondo/blob/master/doc/config.md#namespace-groups

**Comment by borkdude** [comment]
A good example where you need a distinction between `clojure.core/...` and `cljs.core/...` is `:type-mismatch`:

For cljs.core you sometimes need different type signatures, e.g. for `throw`, that you can override as follows:

```
:type-mismatch {:namespaces {cljs.core {throw {:arities {1 {:args [:any]}}}}}}
```

**Comment by vemv** [comment]
(apologies for the radio silence - will get back to this as I clear my backlog)

**Comment by borkdude** [comment]
No worries, I won't have much time this week anyway as I'm heading out to Strange Loop tomorrow

**Comment by borkdude** [comment]
I looked again at this issue and this would be my current recommendation still:

https://github.com/clj-kondo/clj-kondo/issues/2169#issuecomment-1702498380
